### PR TITLE
Improve rope

### LINKED ISSUE #1699: LLaMA 3.1 RoPE
Here's a corrected version of your message:

Dear LitGPT Maintainer,

Thank you for your great work. I encountered an issue while trying to fine-tune LLaMA 3.1 and came here for reference. I was looking for the LLaMA 3.1 RoPE function change, but I couldn't find it in your repository. Based on this PR, it seems like it hasn't been added yet.

https://github.com/Lightning-AI/litgpt/pull/1619/files#diff-3b8a58a4d021803b3171b886bb9162fd659e671131f3f61036f9210cb5d0bc7c

Reference: https://github.com/huggingface/transformers/blob/5c1027bf09717f664b579e01cbb8ec3ef5aeb140/src/transformers/modeling_rope_utils.py#L329-L347

Thanks for your help.

### LINKED ISSUE #1713: Llama 3.1 RoPE not implemented -- Finetuning will lead to spelling/grammar mistakes
### Bug description

I'm re-posting this as the original issue created with a question tag is not receiving attention.

Per https://github.com/Lightning-AI/litgpt/issues/1699

It seems that Llama 3.1 RoPE is not implemented. The Llama 3.1 RoPE is non-standard, according to https://github.com/huggingface/transformers/blob/5c1027bf09717f664b579e01cbb8ec3ef5aeb140/src/transformers/modeling_rope_utils.py#L329-L347

Finetuning will lead to spelling/grammar mistakes.

From @zzhhjjj:

Dear LitGPT Maintainer,

> Thank you for your great work. I encountered an issue while trying to fine-tune LLaMA 3.1 and came here for reference. I was looking for the LLaMA 3.1 RoPE function change, but I couldn't find it in your repository. Based on this PR, it seems like it hasn't been added yet.
> 
> https://github.com/Lightning-AI/litgpt/pull/1619/files#diff-3b8a58a4d021803b3171b886bb9162fd659e671131f3f61036f9210cb5d0bc7c
> 
> Reference: https://github.com/huggingface/transformers/blob/5c1027bf09717f664b579e01cbb8ec3ef5aeb140/src/transformers/modeling_rope_utils.py#L329-L347
> 
> Thanks for your help.

### What operating system are you using?

Linux

### LitGPT Version

``` 
Version: 0.4.11
```

### PR DESCRIPTION
**Pull request by rasbt**
Investigating the RoPE implementation

Fixes #1713 
Fixes #1699

### CODE REVIEW

File: litgpt/model.py [hunk 1]
```diff
@@ -410,17 +437,67 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        wavelen = 2 * torch.pi / theta
+
+        # Initialize adjusted inverse frequencies
+        adjusted_theta = theta.clone()
+
+        # Low Frequency Region: wavelen > low_freq_wavelen
+        mask_low_freq = wavelen > low_freq_wavelen
+        adjusted_theta[mask_low_freq] = theta[mask_low_freq] / factor
+
+        # Medium Frequency Region: high_freq_wavelen ≤ wavelen ≤ low_freq_wavelen
```
Comment (Andrei-Aksionov) [inline litgpt/model.py:489]:
```suggestion
        # Medium Frequency Region: low_freq_wavelen ≤ wavelen ≤ high_freq_wavelen
```

Just assuming by the name - medium frequency region.
Which means that the mask also might be incorrect.

Comment (rasbt) [reply litgpt/model.py:489]: Thanks for the comment! I always appreciate a second pair of eyes! Regarding this particular suggestion:

> # Medium Frequency Region: low_freq_wavelen ≤ wavelen ≤ high_freq_wavelen

I think this is impossible.

Let's assume we have
- `high_freq_wavelen=10`
- `low_freq_wavelen=100`

How can this condition be met?

`100 ≤ wavelen ≤ 10` would not be possible, or am I misunderstanding?

File: litgpt/model.py [hunk 2]
```diff
@@ -108,12 +108,39 @@ def from_name(cls, name: str, **kwargs: Any) -> Self:
...
+                extra_config = {
+                    "original_max_seq_len": self.config.rope_adjustments["original_max_seq_len"],
+                    "factor": self.config.rope_adjustments["factor"],
+                    "low_freq_factor": self.config.rope_adjustments["low_freq_factor"],
+                    "high_freq_factor": self.config.rope_adjustments["high_freq_factor"],
+                }
+            else:
+                # Some but not all parameters are specified; raise an error
+                raise ValueError(
+                    "The following adjusted RoPE parameters are missing in rope_adjustments."
```
[the remark this answers is not in the document]
Comment (rasbt) [reply litgpt/model.py:133]: Good point! I am addressing it here: #1781

File: litgpt/model.py (hunk 1 above)
Comment (Andrei-Aksionov) [reply litgpt/model.py:489]: Oops.
Sometimes I can surprise even myself with my "intelligence".
Who would have thought that in order to have a higher frequency you need a shorter wavelength. 😂

Comment (rasbt) [reply litgpt/model.py:489]: Ohhhh, yeah, well, it's one of these confusing things in physics. But phew! And I was worried that the implementation was wrong haha (but there should be a unit test with the transformer library)
